nixos/modules/services/web-apps/firefly-iii.nix

```diff
@@ -0,0 +1,356 @@
+
+let
+  inherit (lib) mkIf mkOption mkEnableOption literalExpression mdDoc;
+  inherit (lib.types) attrsOf oneOf str int bool path package enum submodule;
+
+  cfg = config.services.firefly-iii;
+  format = pkgs.formats.ini { };
+
+  user = cfg.user;
+  group = cfg.group;
+
+  defaultUser = "firefly-iii";
+  defaultGroup = defaultUser;
+
+  firefly-pkg = pkgs.firefly-iii.override {
+    dataDir = cfg.dataDir;
+    settings = cfg.settings;
+  };
+
+  firefly-php = "${cfg.package.phpPackage}/bin/php";
+  artisan = "${cfg.package}/artisan";
+
+  fileenv-func = ''
+    file_env() {
+      local var="$1"
+      local fileVar="''${var}_FILE"
+      local def="''${2:-}"
+      if [ "''${!var:-}" ] && [ "''${!fileVar:-}" ]; then
+        echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
+      exit 1
+      fi
+      local val="$def"
+      if [ "''${!var:-}" ]; then
+        val="''${!var}"
+      elif [ "''${!fileVar:-}" ]; then
+        val="$(< "''${!fileVar}")"
+      fi
+      printf "$var=\"$val\"\n" >> /tmp/firefly-iii-env.conf
+      export "$var"="$val"
+      unset "$fileVar"
+    }
+
+    envs=(
+      SITE_OWNER
+      APP_KEY
+      DB_CONNECTION
+      DB_HOST
+      DB_PORT
+      DB_DATABASE
+      DB_USERNAME
+      DB_PASSWORD
+      PGSQL_SSL_MODE
+      PGSQL_SSL_ROOT_CERT
+      PGSQL_SSL_CERT
+      PGSQL_SSL_KEY
+      PGSQL_SSL_CRL_FILE
+      REDIS_HOST
+      REDIS_PASSWORD
+      REDIS_PORT
+      COOKIE_DOMAIN
+      MAIL_DRIVER
+      MAIL_HOST
+      MAIL_PORT
+      MAIL_FROM
+      MAIL_USERNAME
+      MAIL_PASSWORD
+      MAIL_ENCRYPTION
+      MAILGUN_DOMAIN
+      MAILGUN_SECRET
+      MAILGUN_ENDPOINT
+      MANDRILL_SECRET
+      SPARKPOST_SECRET
+      MAPBOX_API_KEY
+      FIXER_API_KEY
+      LOGIN_PROVIDER
+      ADLDAP_CONNECTION_SCHEME
+      ADLDAP_CONTROLLERS
+      ADLDAP_PORT
+      ADLDAP_BASEDN
+      ADLDAP_ADMIN_USERNAME
+      ADLDAP_ADMIN_PASSWORD
+      ADLDAP_ACCOUNT_PREFIX
+      ADLDAP_ACCOUNT_SUFFIX
+      WINDOWS_SSO_ENABLED
+      WINDOWS_SSO_DISCOVER
+      WINDOWS_SSO_KEY
+      ADLDAP_SYNC_FIELD
+      TRACKER_SITE_ID
+      TRACKER_URL
+    )
+
+    for e in "''${envs[@]}"; do
+      file_env "$e"
+    done
+  '';
+
+  firefly-iii-maintenance = pkgs.writeScript "firefly-iii-maintenance.sh" ''
+    #!${pkgs.runtimeShell}
+
+    source ${cfg.package}/.env
+
+    ${fileenv-func}
+
+    ${firefly-php} ${artisan} firefly-iii:create-database
+    ${firefly-php} ${artisan} migrate --seed --no-interaction --force
+    ${firefly-php} ${artisan} firefly-iii:decrypt-all
+    ${firefly-php} ${artisan} cache:clear
+    ${firefly-php} ${artisan} firefly-iii:upgrade-database
+    ${firefly-php} ${artisan} firefly-iii:correct-database
+    ${firefly-php} ${artisan} firefly-iii:report-integrity
+    ${firefly-php} ${artisan} firefly-iii:laravel-passport-keys
+    ${firefly-php} ${artisan} cache:clear
+
+    mv /tmp/firefly-iii-env.conf /run/phpfpm/firefly-iii-env.conf
+
+    exit 0
```

Comment:
Since we don't set -e this is really dangerous as it masks failures but even when using set -e, we don't really need this